libr/core/disasm.c

```diff
@@ -6583,3 +6583,186 @@ R_API int r_core_disasm_pdi(RCore *core, int nb_opcodes, int nb_bytes, int fmt)
 	r_core_seek (core, old_offset, true);
 	return err;
 }
+
+R_API bool r_core_disasm_pde(RCore *core, int nb_opcodes, int mode) {
+	PJ *pj = NULL;
+	if (mode == R_MODE_JSON) {
+		pj = pj_new ();
+		if (!pj) {
+			return false;
+		}
+		pj_a (pj);
+	}
+	if (!core->anal->esil) {
+		r_core_cmd0 (core, "aei");
+		if (!r_config_get_i (core->config, "cfg.debug")) {
+			r_core_cmd0 (core, "aeim");
+		}
+	}
+	const ut64 ocoffset = core->offset;
+	const ut64 ocbsize = core->blocksize;
+	RAnalEsil *esil = core->anal->esil;
+	RReg *reg = core->anal->reg;
+	RList *ocache = core->io->cache;
+	RCache *ocacheb = core->io->buffer;
+	const int ocached = core->io->cached;
+	const bool ocachec = r_config_get_i (core->config, "io.cache");
+	const bool ojmp_lines = r_config_get_i (core->config, "asm.lines");
+	const bool hide_branches = r_config_get_i (core->config, "asm.followcf.hide");
+	if (ocacheb && ocache) {
+		RCache *c = R_NEW0 (RCache);
+		c->base = ocacheb->base;
+		if (ocacheb->len) {
+			ut8 *buf = malloc (ocacheb->len);
+			if (!buf) {
+				pj_free (pj);
+				r_cache_free (c);
+				return false;
+			}
+			memcpy (buf, ocacheb->buf, ocacheb->len);
+			c->buf = buf;
+			c->len = ocacheb->len;
+		}
+		core->io->buffer = c;
+		core->io->cache = r_list_clone (ocache);
+	} else {
+		r_io_cache_init (core->io);
+	}
+	r_reg_arena_push (reg);
+	r_config_set_i (core->config, "io.cache", 1);
+	r_config_set_i (core->config, "asm.lines", 0);
+	const ut64 read_len = 32;
+	size_t buf_sz = 0x100, block_sz = 0, block_instr = 0;
+	RRegItem *pc = r_reg_get (reg, "PC", R_REG_TYPE_ALL);
+	ut64 block_start = r_reg_get_value (reg, pc);
+	ut8 *buf = malloc (buf_sz);
+	int i;
+	for (i = 0; i < nb_opcodes; i++) {
+		const ut64 op_addr = r_reg_get_value (reg, pc);
+		if (block_sz + read_len > buf_sz) {
```

Comment:
move all this logic of checking buf size and read size and contiguous reads stuff into a separate function. or just use RBuffer which may do that with the io backend.